File: R/PCAproj.R

PCAproj <- function (x, k = 2, method = c ("mad", "sd", "qn"), CalcMethod = c("eachobs", "lincomb", "sphere"), nmax = 1000, update = TRUE, scores = TRUE, maxit = 5, maxhalf = 5, scale = NULL, center = l1median_NLM, zero.tol = 1e-16, control)
{
	if (!missing (control))
		###ParseControlStructure (control, c("k", "method", "CalcMethod", "nmax", "update", "scores", "maxit", "maxhalf"))
		.ParseControlStructure (control, c("k", "method", "CalcMethod", "nmax", "update", "scores", "maxit", "maxhalf", "center", "scale", "zero.tol"))

	method <- .getScaleMethod (method)

	CalcMethod <- match.arg (CalcMethod[1], c("eachobs", "lincomb", "sphere"))

	x <- .Conv2Matrix (x, substitute (x))

	n = nrow (x)
	p = ncol (x)

	if( k > min(n,p))
		stop ('k too large')

	if(p > n)
	{
			svdx = svd(t(x))
			x = svdx$v %*% diag(svdx$d)
			pold=p
			p=n
	} else
			pold=p

	DataObj = ScaleAdv (x, scale = scale, center = center)

    if (pold > n) # center and scale must have original data dimension:
	{
		DataObj$center <- as.vector(svdx$u%*%DataObj$center)
		DataObj$scale <- ScaleAdv(x%*%t(svdx$u),center=NULL,scale=scale)$scale
	}

	y = DataObj$x

#	m = l1median(x)
#	y = t(t(x) - m)

	if (scores)
		scoresize <- n * k
	else
		scoresize <- 0

	if (CalcMethod == "lincomb")
	{
		update = FALSE
		if (nmax > n)
		{
			aux = matrix (runif ((nmax-n) * n), nrow = nmax-n)
			y = rbind (y, t(t(aux %*% as.matrix(x)) - DataObj$center))
		}
	}
	else if (CalcMethod == "sphere")
	{
		update = FALSE
		if(nmax  >n)
			#y[(n+1):nmax,] =	rmvnorm(nmax-n, rep(0,p), diag (p))
			y = rbind (y,		rmvnorm(nmax-n, rep(0,p), diag (p)))
	}

	nn = nrow (y)

	if (update)
		ret.C = .C ("pcaProj_up", PACKAGE="pcaPP"
				, as.integer (c(nn, p, n, k, method, scores, maxit, maxhalf))
				, as.double (zero.tol)
				, as.double (y)
				, scores = double (scoresize)
				, loadings = double (p * k)
				, lambda = double (k)
			)
	else
		ret.C = .C ("pcaProj",  PACKAGE="pcaPP"
				, as.integer (c(nn, p, n, k, method, scores))
				, as.double (zero.tol)
				, as.double (y)
				, scores = double (scoresize)
				, loadings = double (p * k)
				, lambda = double (k)
			)


	veig = matrix (ret.C$loadings, ncol = k)

   if(pold>n)
		veig = svdx$u %*% veig

	if (scores)
		.DataPostProc (DataObj, ret.C$lambda, veig, matrix (ret.C$scores, ncol = k) , match.call(), scores)
	else
		.DataPostProc (DataObj, ret.C$lambda, veig, NULL, match.call(), scores)
}

File: matlab/PCAproj.m
function [ ret ] = PCAproj (x, k, method, CalcMethod, nmax, update, scores, maxit, maxhalf, scale, center, zero_tol)

    if (nargin < 12)
        zero_tol = 1e-16 ;
        if (nargin < 11)
            center = 'l1median_HoCr' ;
            if (nargin < 10)
                scale = null (1) ;
                if (nargin < 9)
                    maxhalf = 5 ;
                    if (nargin < 8)
                        maxit = 5 ;
                        if (nargin < 7)
                            scores = true ;
                            if (nargin < 6)
                                update = true ;
                                if (nargin < 5)
                                    nmax = 1000 ;
                                    if (nargin < 4)
                                        CalcMethod = 'eachobs' ;
                                        if (nargin < 3)
                                            method = 'mad' ;
                                            if (nargin < 2)
                                                k = 2 ;
                                                if (nargin < 1)
                                                    error ('Not enough input arguments provided.') ;
                                                end
                                            end
                                        end
                                    end
                                end
                            end
                        end
                    end
                end
            end
        end
    end
    
	method = p_GetScaleMethod (method) ;

	[n, p] = size (x) ;

	if( k > min(n,p))
		error ('k too large')
    end

	if(p > n)
        
            [svdx.u, S, svdx.v] = svd(x') ;

            svdx.u = svdx.u (:, 1:size(x,1)) ;
            svdx.d = diag (S) ;
    		x = svdx.v * diag (svdx.d) ;

%			svdx = svd(t(x)) ;
%			x = svdx.v * diag(svdx.d) ;

			pold = p ;
			p = n ;
	else
			pold = p ;
    end

	DataObj = p_ScaleAdv (x, center, scale) ;

    if (pold > n) % center and scale must have original data dimension:
		DataObj.center = svdx.u * DataObj.center' ;
        scaled = p_ScaleAdv (x * svdx.u', null (1), scale) ;
		DataObj.scale = scaled.scale ;
    end

	y = DataObj.x ;

    if (scores)
		scoresize = n * k ;
	else
		scoresize = 0 ;
    end

	if (strcmp (CalcMethod, 'lincomb'))
        update = false ;
		if (nmax > n)
            aux = rand (nmax - n, n) ;
			%aux = matrix (runif ((nmax-n) * n), nrow = nmax-n) ;
            yadd = (aux * x) - (ones (size (aux, 1), 1) * DataObj.center) ;
            %yadd = (aux * y) ; % shouldn't this be better?
            y = [y ; yadd] ;
			%y = rbind (y, t(t(aux * as.matrix(x)) - DataObj.center)) ;
        end
    elseif (strcmp (CalcMethod, 'sphere'))
        update = false ;
		if(nmax > n)
            yadd = randn (nmax - n, p) ;
            y = [y ; yadd] ;
			% y = rbind (y, randn (nmax-n, p)) ; %rmvnorm(nmax-n, rep(0,p), diag (p))) ;
        end
    elseif  (strcmp (CalcMethod, 'eachobs'))
    else
       error ('Argument CalcMethod: invalid value.')
    end

	nn = size (y, 1) ;

	if (update)
        nParIn = [n, k, method, scores, maxit, maxhalf] ;
        [z, l, s] = pcaPP (6, nParIn, zero_tol, y) ;
    else
        nParIn = [n, k, method, scores] ;
        [z, l, s] = pcaPP (5, nParIn, zero_tol, y) ;
    end

	%veig = matrix (ret.C.loadings, ncol = k) ;

    if (pold > n)
		l = svdx.u * l ;
    end

	if (scores)
		ret = p_DataPostProc (DataObj, s, l, z(1:n, :) , scores) ;
	else
		ret = p_DataPostProc (DataObj, s, l, null (1), scores) ;
    end
